### ISSUE TITLE
[question] conan upload: how to choose packages without user/channel

### ISSUE THREAD

**Issue by cbecker** [conan-io/conan#7967]
I want to upload recipes from the `_/_` user/channel with `conan upload`, but I'm not sure how to do this with pattern matching.

If I try
```
conan upload "*/*@"
```
it selects packages with any user/channel. I also tried
```
conan upload "*/*@_/_"
```
but in that case I get no matching packages. Is there something I'm missing regarding the pattern matching expressions used by conan upload? thanks

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by cbecker** [comment]
after further investigation, it looks like it's not possible with the current implementation, because of `_partial_match(pattern, reference)` in `search.py`.

Do you think having a flag to negate the pattern would be useful for such cases?

**Comment by memsharded** [comment]
Hi @cbecker 

Yes, you are right, it seems right now it is not possible to select those in that way.

We are certainly going to change the command line (@czoido) which will probably mean that unless adding explictly "*" it will not be a pattern, so that will solve the issue. The only thing is that it seems it cannot be done now without breaking.

If what you are doing is caching a copy of packages without user and channel download from ConanCenter, have you tried proxying ConanCenter with a remote repository in Artifactory? That will do it automatically for you.

**Comment by cbecker** [comment]
Thanks @memsharded . Indeed, I was trying to simulate our own conan-center, but with specific versions instead of what conan-center offers online at the moment of calling conan (https://github.com/conan-io/conan-center-index/issues/3077)

So we have our own conan server, where we push `conan/stable` and `_/_` from all recipes in a given commit hash of `conan-center-index`. I don't like much the idea but it keeps our build intact and we can update to the latest conan-center-index when we have time available and we can fix the issues, and like that iterate.

What I ended up doing is something that is much slower but works:

```
conan search --raw | grep -v '@' | xargs -L 1 -P $(( $(nproc) * 2 )) conan upload -r <remote> -c
```

**Comment by memsharded** [comment]
> So we have our own conan server, where we push conan/stable and _/_ from all recipes in a given commit hash of conan-center-index. I don't like much the idea but it keeps our build intact and we can update to the latest conan-center-index when we have time available and we can fix the issues, and like that iterate.

No, it is not a bad idea, it makes sense to have your own fork, and control 100% of your third parties. Actually it is a good practice, and using directly a public central repository in production is not recommended, and I am not talking about Conan but also about NPM and others (remember the NPM leftpad...). There are different mechanisms to isolate yourself from problems of third party repos, and hosting your own server is the best one. Then it can be populated caching the central one or directly by the user. In Conan, due to corporate and industry policies, many users cannot use the binaries from ConanCenter, so they need to build their own binaries and we are very happy we that.

Do you have many dependencies from ConanCenter, if not too many, you can workaround by hardcoding them in the CI or in a script or by manually filtering y/n in the upload? If you have many, and they change and it is too painful, let us know, we can't promise but maybe we can try some other alternative in next Conan 1.32

**Comment by cbecker** [comment]
Thanks for the reply, I'm glad to hear that it's not a crazy idea and that it also makes sense to other projects/companies.

We have probably around 15 dependencies from conan-center. The thing is that conan center is great (so big thanks for taking care of this!) in the sense that there are so many useful recipes there. So having them all available as recipes (without binaries) also proves useful.

Maybe a more important point: build dependencies can be an issue and that was the 1st reason why we opted to push all the recipes to our remote. Then no more surprises if we need to build from scratch on a new distro/gcc/libc, etc.

**Comment by memsharded** [comment]
Hi @cbecker 

As it seems you have a workaround, or as you have like 15 packages from ConanCenter, explicitly listing them (in CI or in a script) might be doable, I am moving this to Conan 2.0. The issue is that it seems impossible to do it without breaking, or without requiring a hacky solution, like a configuration flag, or the negate patterns. So if we could wait until we have the right implementation in which ``*@`` only matches packages without user/channel, that would be great.

**Comment by memsharded** [comment]
This has been fixed in the patterns of 2.0, not only upload but others. A pattern that ends in ``@`` will mean no user channel. So ``conan upload zlib/1.2.*`` might match all user and channels, but ``conan upload zlib/1.2.*@`` will assume no user and channel matches only
